### PR TITLE
nixos/stage-1: colon-separated multi-device support

### LINKED ISSUE #152569: Bcachefs multi-device stage-1 waiting
### Describe the bug
``/dev/sdXX:/dev/sdXX:/dev/sdXX`` is obviously not a valid device in most cases, so it times out and then mounts anyway.
Ideally, we would skip the timer for bcachefs and just mount anyway.

### Steps To Reproduce
1. Use a multi-device bcachefs root

### Expected behavior
Not having to wait for timeout to mount.

### Additional context
Likely need to add a condition check if the mount/unlock we're waiting on is a colon-seperated line and/or bcachefs, and just skip waiting. (eg: ``/dev/sda1:/dev/sda2:/dev/sda3``)
https://github.com/NixOS/nixpkgs/blob/3239e947d1bfb7cfef6fdc1a88753059d4847aaa/nixos/modules/system/boot/stage-1-init.sh#L85

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.11, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(mado): `"home-manager-21.11, nixos-21.11.334821.d887ac7aee9, nixos-small-21.11.334860.4fe043fcc2e"`
 - channels(root): `"nixos-21.11.334821.d887ac7aee9, home-manager-21.11, nixos-small-21.11.334860.4fe043fcc2e"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by Madouura**
###### Motivation for this change
Fixes #152569

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Madouura** [comment]
cc @Mic92

**Comment by Mic92** [comment]
Just for clarification, with bcachefs you do something like?

```
{
  fileSystems."/" = {
    device = "/dev/sda1:/dev/sda2:/dev/sda3";
    fsType = "bcachefs";
  };
}

**Comment by Madouura** [comment]
Yes, this is from my own hardware-configuration.nix, generated by ``nixos-generate-config``.
```nix
fileSystems."/" =
  { device = "/dev/sda1:/dev/nvme0n1p2:/dev/nvme1n1p1";
    fsType = "bcachefs";
  };
```

File: nixos/modules/system/boot/stage-1-init.sh [hunk 1]
```diff
@@ -81,30 +81,33 @@ ln -s /proc/mounts /etc/mtab # to shut up mke2fs
 touch /etc/udev/hwdb.bin # to shut up udev
 touch /etc/initrd-release
 
-# Function for waiting a device to appear.
+# Function for waiting for device(s) to appear.
 waitDevice() {
     local device="$1"
+    local IFS=':'
```
Comment (Mic92) [inline nixos/modules/system/boot/stage-1-init.sh:87]:
```suggestion
    # Split device string using : as a delimiter as bcachefs uses this for multi-device filesystems i.e. /dev/sda1:/dev/sda2:/dev/sda3
    local IFS=':'
```

**Comment by Mic92** [comment]
It would be great if we could get installer nixos test in future that would test this for bcachefs....

File: nixos/modules/system/boot/stage-1-init.sh (hunk 1 above)
Comment (Madouura) [reply nixos/modules/system/boot/stage-1-init.sh:89]: Done!

**Comment by Madouura** [comment]
I agree, only way I was able to test if this worked for single-device systems was a vm.nix build.

**Comment by Madouura** [comment]
Would that be possible in https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/installer.nix?

**Comment by Madouura** [comment]
Looks like we might be able to, as the kernel can be changed to bcachefs. (https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/installer.nix#L374)
Should I use this PR for the tests or make a new one once this is merged?
Seems easy enough.

**Comment by Mic92** [comment]
> Looks like we might be able to, as the kernel can be changed to bcachefs. (https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/installer.nix#L374) Should I use this PR for the tests or make a new one once this is merged? Seems easy enough.

Let's make a new PR for this.

**Comment by Madouura** [comment]
@dasJ Good for merge?

**Comment by dasJ** [comment]
Just to make sure - are you planning to wait for all devices or just one? All, right?

**Comment by Madouura** [comment]
All devices, yeah.
Would be kind of iffy if we just waited for one and assumed the rest were fine.

**Pull request merged**